src/tools/illink/src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2016,7 +2018,9 @@ internal void MarkStaticConstructorVisibleToReflection (TypeDefinition type, in
 			if (type.Scope is ModuleDefinition module)
 				MarkModule (module, new DependencyInfo (DependencyKind.ScopeOfType, type));
 
-			_dependencyGraph.AddRoot (_nodeFactory.GetTypeNode (type), Enum.GetName (reason.Kind));
+			var typeNode = _nodeFactory.GetTypeNode (type);
+			var rootNode = _nodeFactory.GetRootTracingNode (typeNode, NodeFactory.DependencyKindToStringMap[reason.Kind], reason.Source);
```

Comment:
To avoid the RootTracingNode, I'll make two implementations of MarkType/Method/etc. One will be for calls from outside of the graph that will mark the object in the AnnotationStore with the reason/dependency to log it to the trimmer's dependency logger. The other will mark it without a reason and return a Node to be added to the graph, then the EdgeVisitor will log the edges to the trimmer's dependency logger.

Reply:
Ended up just using an optional bool parameter in MarkType rather than two definitions. I'll solve it in a nicer way in a follow up, and do the same for MarkMethod once we return a MethodNode from GetStaticDependencies, but I wanted to keep this PR more targeted.